File: flowersec-go/internal/defaults/defaults_contract_test.go
package defaults_test

import (
	"encoding/json"
	"os"
	"path/filepath"
	"runtime"
	"testing"
	"time"

	"github.com/floegence/flowersec/flowersec-go/internal/defaults"
)

func TestDefaultsMatchStabilityContract(t *testing.T) {
	_, filename, _, ok := runtime.Caller(0)
	if !ok {
		t.Fatal("resolve test path")
	}
	path := filepath.Join(filepath.Dir(filename), "..", "..", "..", "stability", "sdk_defaults.json")
	data, err := os.ReadFile(path)
	if err != nil {
		t.Fatal(err)
	}
	var manifest struct {
		Transport    map[string]float64 `json:"transport"`
		E2EE         map[string]float64 `json:"e2ee"`
		Yamux        map[string]float64 `json:"yamux"`
		RPC          map[string]float64 `json:"rpc"`
		Controlplane map[string]float64 `json:"controlplane"`
		Proxy        map[string]float64 `json:"proxy"`
		Reconnect    map[string]float64 `json:"reconnect"`
	}
	if err := json.Unmarshal(data, &manifest); err != nil {
		t.Fatal(err)
	}

	assertDurationMS(t, defaults.ConnectTimeout, manifest.Transport["connect_timeout_ms"])
	assertDurationMS(t, defaults.HandshakeTimeout, manifest.Transport["handshake_timeout_ms"])
	assertDurationMS(t, defaults.HandshakeClockSkew, manifest.Transport["handshake_clock_skew_ms"])
	assertInt(t, defaults.MaxHandshakePayloadBytes, manifest.E2EE["max_handshake_payload_bytes"])
	assertInt(t, defaults.MaxRecordBytes, manifest.E2EE["max_record_bytes"])
	assertInt(t, defaults.OutboundRecordChunkBytes, manifest.E2EE["outbound_record_chunk_bytes"])
	assertInt(t, defaults.MaxInboundBufferedBytes, manifest.E2EE["max_inbound_buffered_bytes"])
	assertInt(t, defaults.MaxOutboundBufferedBytes, manifest.E2EE["max_outbound_buffered_bytes"])
	assertInt(t, defaults.YamuxMaxActiveStreams, manifest.Yamux["max_active_streams"])
	assertInt(t, defaults.YamuxMaxInboundStreams, manifest.Yamux["max_inbound_streams"])
	assertInt(t, defaults.YamuxMaxFrameBytes, manifest.Yamux["max_frame_bytes"])
	assertInt(t, defaults.YamuxPreferredOutboundFrameBytes, manifest.Yamux["preferred_outbound_frame_bytes"])
	assertInt(t, defaults.YamuxMaxStreamWriteQueueBytes, manifest.Yamux["max_stream_write_queue_bytes"])
	assertInt(t, defaults.YamuxMaxStreamReceiveBytes, manifest.Yamux["max_stream_receive_bytes"])
	assertInt(t, defaults.YamuxMaxSessionReceiveBytes, manifest.Yamux["max_session_receive_bytes"])
	assertInt(t, defaults.RPCMaxJSONFrameBytes, manifest.RPC["max_json_frame_bytes"])
	assertInt(t, defaults.RPCMaxConcurrentRequests, manifest.RPC["max_concurrent_requests"])
	assertInt(t, defaults.RPCMaxQueuedRequests, manifest.RPC["max_queued_requests"])
	assertInt(t, defaults.RPCMaxQueuedNotifications, manifest.RPC["max_queued_notifications"])
	assertInt(t, defaults.ControlplaneMaxRequestBodyBytes, manifest.Controlplane["max_request_body_bytes"])
	assertInt(t, defaults.ControlplaneMaxResponseBodyBytes, manifest.Controlplane["max_response_body_bytes"])
	assertInt(t, defaults.ProxyMaxJSONFrameBytes, manifest.Proxy["max_json_frame_bytes"])
	assertInt(t, defaults.ProxyMaxChunkBytes, manifest.Proxy["max_chunk_bytes"])
	assertInt(t, defaults.ProxyMaxBodyBytes, manifest.Proxy["max_body_bytes"])
	assertInt(t, defaults.ProxyMaxWSFrameBytes, manifest.Proxy["max_ws_frame_bytes"])
	assertDurationMS(t, defaults.ProxyDefaultTimeout, manifest.Proxy["default_timeout_ms"])
	assertDurationMS(t, defaults.ProxyMaxTimeout, manifest.Proxy["max_timeout_ms"])
	assertInt(t, defaults.ReconnectMaxAttempts, manifest.Reconnect["max_attempts"])
	assertDurationMS(t, defaults.ReconnectInitialDelay, manifest.Reconnect["initial_delay_ms"])
	assertDurationMS(t, defaults.ReconnectMaxDelay, manifest.Reconnect["max_delay_ms"])
	if defaults.ReconnectFactor != manifest.Reconnect["factor"] || defaults.ReconnectJitterRatio != manifest.Reconnect["jitter_ratio"] {
		t.Fatal("reconnect floating-point defaults do not match stability contract")
	}
}

func assertDurationMS(t *testing.T, got time.Duration, want float64) {
	t.Helper()
	if float64(got.Milliseconds()) != want {
		t.Fatalf("duration default = %dms, want %.0fms", got.Milliseconds(), want)
	}
}

func assertInt(t *testing.T, got int, want float64) {
	t.Helper()
	if float64(got) != want {
		t.Fatalf("integer default = %d, want %.0f", got, want)
	}
}

File: flowersec-ts/src/defaults_contract.test.ts
import { readFile } from "node:fs/promises";
import { describe, expect, test } from "vitest";
import { SDK_DEFAULTS } from "./defaults.js";

describe("SDK defaults contract", () => {
  test("matches the shared stability manifest", async () => {
    const path = new URL("../../stability/sdk_defaults.json", import.meta.url);
    const manifest = JSON.parse(await readFile(path, "utf8")) as Record<string, Record<string, number>>;

    expect(SDK_DEFAULTS.transport).toEqual({
      connectTimeoutMs: manifest.transport!.connect_timeout_ms,
      handshakeTimeoutMs: manifest.transport!.handshake_timeout_ms,
      handshakeClockSkewMs: manifest.transport!.handshake_clock_skew_ms,
    });
    expect(SDK_DEFAULTS.e2ee).toEqual({
      maxHandshakePayloadBytes: manifest.e2ee!.max_handshake_payload_bytes,
      maxRecordBytes: manifest.e2ee!.max_record_bytes,
      outboundRecordChunkBytes: manifest.e2ee!.outbound_record_chunk_bytes,
      maxInboundBufferedBytes: manifest.e2ee!.max_inbound_buffered_bytes,
      maxOutboundBufferedBytes: manifest.e2ee!.max_outbound_buffered_bytes,
    });
    expect(SDK_DEFAULTS.yamux).toEqual({
      maxActiveStreams: manifest.yamux!.max_active_streams,
      maxInboundStreams: manifest.yamux!.max_inbound_streams,
      maxFrameBytes: manifest.yamux!.max_frame_bytes,
      preferredOutboundFrameBytes: manifest.yamux!.preferred_outbound_frame_bytes,
      maxStreamWriteQueueBytes: manifest.yamux!.max_stream_write_queue_bytes,
      maxStreamReceiveBytes: manifest.yamux!.max_stream_receive_bytes,
      maxSessionReceiveBytes: manifest.yamux!.max_session_receive_bytes,
    });
    expect(SDK_DEFAULTS.rpc).toEqual({
      maxJsonFrameBytes: manifest.rpc!.max_json_frame_bytes,
      maxConcurrentRequests: manifest.rpc!.max_concurrent_requests,
      maxQueuedRequests: manifest.rpc!.max_queued_requests,
      maxQueuedNotifications: manifest.rpc!.max_queued_notifications,
    });
    expect(SDK_DEFAULTS.controlplane).toEqual({
      maxRequestBodyBytes: manifest.controlplane!.max_request_body_bytes,
      maxResponseBodyBytes: manifest.controlplane!.max_response_body_bytes,
    });
    expect(SDK_DEFAULTS.proxy).toEqual({
      maxJsonFrameBytes: manifest.proxy!.max_json_frame_bytes,
      maxChunkBytes: manifest.proxy!.max_chunk_bytes,
      maxBodyBytes: manifest.proxy!.max_body_bytes,
      maxWsFrameBytes: manifest.proxy!.max_ws_frame_bytes,
      defaultTimeoutMs: manifest.proxy!.default_timeout_ms,
      maxTimeoutMs: manifest.proxy!.max_timeout_ms,
    });
    expect(SDK_DEFAULTS.reconnect).toEqual({
      maxAttempts: manifest.reconnect!.max_attempts,
      initialDelayMs: manifest.reconnect!.initial_delay_ms,
      maxDelayMs: manifest.reconnect!.max_delay_ms,
      factor: manifest.reconnect!.factor,
      jitterRatio: manifest.reconnect!.jitter_ratio,
    });
  });
});
